File: scripts/migrate_env.py
#!/usr/bin/env python3
"""Migrate non-secret settings from .env to ~/.atrophy/config.json.

Reads the project .env file, identifies settings vs secrets, moves settings
into config.json, and leaves only secrets in .env.

Safe to run multiple times — only migrates keys not already in config.json.
"""
import json
import re
import sys
from pathlib import Path

PROJECT_ROOT = Path(__file__).parent.parent
ENV_PATH = PROJECT_ROOT / ".env"
CONFIG_PATH = Path.home() / ".atrophy" / "config.json"

# Keys that are secrets and should STAY in .env
SECRET_KEYS = {
    "ELEVENLABS_API_KEY",
    "FAL_KEY",
    "TELEGRAM_BOT_TOKEN",
    "OPENAI_API_KEY",
    "ANTHROPIC_API_KEY",
}

# Keys that look like secrets (heuristic)
SECRET_PATTERNS = [
    re.compile(r".*_KEY$"),
    re.compile(r".*_SECRET$"),
    re.compile(r".*_TOKEN$"),
    re.compile(r".*_PASSWORD$"),
]


def is_secret(key: str) -> bool:
    if key in SECRET_KEYS:
        return True
    return any(p.match(key) for p in SECRET_PATTERNS)


def parse_env(path: Path) -> list[tuple[str, str, str]]:
    """Parse .env file. Returns list of (key, value, raw_line)."""
    entries = []
    if not path.exists():
        return entries
    for line in path.read_text().splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            entries.append(("", "", line))
            continue
        match = re.match(r"^([A-Za-z_][A-Za-z0-9_]*)=(.*)$", stripped)
        if match:
            entries.append((match.group(1), match.group(2), line))
        else:
            entries.append(("", "", line))
    return entries


def main():
    if not ENV_PATH.exists():
        print("No .env file found — nothing to migrate.")
        return

    entries = parse_env(ENV_PATH)
    if not entries:
        print(".env is empty — nothing to migrate.")
        return

    # Load existing config.json
    config = {}
    if CONFIG_PATH.exists():
        try:
            config = json.loads(CONFIG_PATH.read_text())
        except (json.JSONDecodeError, OSError):
            config = {}

    migrated = []
    secrets_kept = []
    remaining_lines = []

    for key, value, raw_line in entries:
        if not key:
            remaining_lines.append(raw_line)
            continue

        if is_secret(key):
            secrets_kept.append(key)
            remaining_lines.append(raw_line)
            continue

        # Non-secret setting — migrate to config.json if not already there
        if key not in config:
            config[key] = value
            migrated.append(key)
            print(f"  Migrated: {key}={value}")
        else:
            print(f"  Skipped (already in config.json): {key}")

    if migrated:
        CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)
        CONFIG_PATH.write_text(json.dumps(config, indent=2) + "\n")
        print(f"\nWrote {len(migrated)} setting(s) to {CONFIG_PATH}")

        # Rewrite .env with only secrets
        ENV_PATH.write_text("\n".join(remaining_lines).strip() + "\n")
        print(f"Updated .env — {len(secrets_kept)} secret(s) remain.")
    else:
        print("\nNothing to migrate — all .env entries are secrets (or already in config.json).")

    if secrets_kept:
        print(f"\nSecrets kept in .env: {', '.join(secrets_kept)}")


if __name__ == "__main__":
    main()

File: src/main/migrate-env.ts
/**
 * Migrate legacy .env format to config.json.
 * Port of scripts/migrate_env.py.
 *
 * Reads ~/.atrophy/.env, parses key=value pairs, identifies secrets vs settings,
 * and writes non-secret settings into ~/.atrophy/config.json. Secrets stay in .env.
 *
 * Safe to run multiple times - only migrates keys not already in config.json.
 */

import * as fs from 'fs';
import * as path from 'path';
import { USER_DATA } from './config';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface MigrationResult {
  /** Keys successfully moved to config.json */
  migrated: string[];
  /** Keys skipped because they already exist in config.json */
  skipped: string[];
  /** Secret keys left in .env */
  secretsKept: string[];
  /** True if config.json was written */
  configWritten: boolean;
  /** True if .env was rewritten */
  envRewritten: boolean;
}

interface EnvEntry {
  key: string;
  value: string;
  rawLine: string;
}

// ---------------------------------------------------------------------------
// Secret detection
// ---------------------------------------------------------------------------

const SECRET_KEYS = new Set([
  'ELEVENLABS_API_KEY',
  'FAL_KEY',
  'TELEGRAM_BOT_TOKEN',
  'OPENAI_API_KEY',
  'ANTHROPIC_API_KEY',
]);

const SECRET_PATTERNS = [
  /_KEY$/,
  /_SECRET$/,
  /_TOKEN$/,
  /_PASSWORD$/,
];

function isSecret(key: string): boolean {
  if (SECRET_KEYS.has(key)) return true;
  return SECRET_PATTERNS.some((p) => p.test(key));
}

// ---------------------------------------------------------------------------
// .env parser
// ---------------------------------------------------------------------------

const ENV_LINE_RE = /^([A-Za-z_][A-Za-z0-9_]*)=(.*)$/;

function parseEnv(filePath: string): EnvEntry[] {
  if (!fs.existsSync(filePath)) return [];

  const text = fs.readFileSync(filePath, 'utf-8');
  const entries: EnvEntry[] = [];

  for (const line of text.split('\n')) {
    const stripped = line.trim();
    if (!stripped || stripped.startsWith('#')) {
      entries.push({ key: '', value: '', rawLine: line });
      continue;
    }
    const match = ENV_LINE_RE.exec(stripped);
    if (match) {
      entries.push({ key: match[1], value: match[2], rawLine: line });
    } else {
      entries.push({ key: '', value: '', rawLine: line });
    }
  }

  return entries;
}

// ---------------------------------------------------------------------------
// Migration
// ---------------------------------------------------------------------------

/**
 * Migrate non-secret settings from ~/.atrophy/.env into ~/.atrophy/config.json.
 *
 * - Reads .env, separates secrets from settings
 * - Writes new settings into config.json (does not overwrite existing keys)
 * - Rewrites .env to contain only secrets and comments
 *
 * Returns a summary of what was done.
 */
export function migrateEnv(): MigrationResult {
  const envPath = path.join(USER_DATA, '.env');
  const configPath = path.join(USER_DATA, 'config.json');

  const result: MigrationResult = {
    migrated: [],
    skipped: [],
    secretsKept: [],
    configWritten: false,
    envRewritten: false,
  };

  const entries = parseEnv(envPath);
  if (entries.length === 0) {
    return result;
  }

  // Load existing config.json
  let config: Record<string, unknown> = {};
  try {
    if (fs.existsSync(configPath)) {
      config = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
    }
  } catch {
    config = {};
  }

  const remainingLines: string[] = [];

  for (const entry of entries) {
    if (!entry.key) {
      remainingLines.push(entry.rawLine);
      continue;
    }

    if (isSecret(entry.key)) {
      result.secretsKept.push(entry.key);
      remainingLines.push(entry.rawLine);
      continue;
    }

    // Non-secret setting - migrate if not already present
    if (!(entry.key in config)) {
      config[entry.key] = entry.value;
      result.migrated.push(entry.key);
    } else {
      result.skipped.push(entry.key);
    }
  }

  if (result.migrated.length > 0) {
    fs.mkdirSync(path.dirname(configPath), { recursive: true });
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2) + '\n', { mode: 0o600 });
    result.configWritten = true;

    // Rewrite .env with only secrets and comments
    const newEnvContent = remainingLines.join('\n').trim() + '\n';
    fs.writeFileSync(envPath, newEnvContent);
    result.envRewritten = true;
  }

  return result;
}
